### PR TITLE
feat(agent):metrics for errors in agent

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [x] I have filled out the description and linked the related issues.

### Description
Adds metrics for errors for Grafana dashboard

#### Motivation and Context (Optional)
An extension of Agent feature in storage incentives. 

### Related Issue (Optional)

NA
### Screenshots (if appropriate):

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3449)

### CODE REVIEW

**Review by notanatol — APPROVED** [review]

**Comment by swarmHaseeb** [comment]
@istae @vladopajic ci errs unrelated to pr

File: pkg/storageincentives/agent.go [hunk 1]
```diff
@@ -119,8 +119,14 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		mtx.Unlock()
 
 		if round-1 == sampleRound { // the sample has to come from previous round to be able to commit it
+			a.metrics.CommitPhase.Inc()
```
Comment (istae) [inline pkg/storageincentives/agent.go:122]: For the phases, we just need a counter and then an error counter,
we do not need this AllPhaseEvents or AllContractCalls

File: pkg/storageincentives/agent.go [hunk 2]
```diff
@@ -228,8 +240,10 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		case <-time.After(blockTime * time.Duration(checkEvery)):
 		}
 
+		a.metrics.BackendCalls.Inc()
```
Comment (istae) [inline pkg/storageincentives/agent.go:243]: nice one

File: pkg/storageincentives/agent.go [hunk 3]
```diff
@@ -308,29 +331,40 @@ func (a *Agent) play(ctx context.Context) (uint8, []byte, error) {
 
 	storageRadius := a.reserve.GetReserveState().StorageRadius
 
+	a.metrics.AllContractCalls.Inc()
 	isPlaying, err := a.contract.IsPlaying(ctx, storageRadius)
 	if !isPlaying || err != nil {
+		a.metrics.AllContractCallErrors.Inc()
+		a.metrics.PhasesErrors.All.Inc()
+		a.metrics.PhasesErrors.NeighborhoodSelected.Inc()
```
Comment (istae) [inline pkg/storageincentives/agent.go:339]: rename to something like `IsPlayingCheck`

File: pkg/storageincentives/metrics.go [hunk 4]
```diff
@@ -18,6 +24,22 @@ type metrics struct {
 	NeighborhoodSelected prometheus.Counter
 	SampleDuration       prometheus.Gauge
 	Round                prometheus.Gauge
+
+	// total calls to chain backend
+	BackendCalls  prometheus.Counter
+	BackendErrors prometheus.Counter
+
+	// metrics for err processing
+	PhasesErrors phasesErrors
```
Comment (istae) [inline pkg/storageincentives/metrics.go:33]: No need for a second layer struct.

File: pkg/storageincentives/agent.go [hunk 5]
```diff
@@ -119,8 +119,14 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		mtx.Unlock()
 
 		if round-1 == sampleRound { // the sample has to come from previous round to be able to commit it
+			a.metrics.CommitPhase.Inc()
+			a.metrics.AllPhaseEvents.Inc()
+			a.metrics.AllContractCalls.Inc()
 			obf, err := a.commit(ctx, storageRadius, reserveSample)
 			if err != nil {
+				a.metrics.PhasesErrors.All.Inc()
```
Comment (istae) [inline pkg/storageincentives/agent.go:127]: a metric for all errors is not needed, just specific ones

File: pkg/storageincentives/agent.go [hunk 6]
```diff
@@ -228,8 +240,10 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		case <-time.After(blockTime * time.Duration(checkEvery)):
 		}
 
+		a.metrics.BackendCalls.Inc()
 		block, err := a.backend.BlockNumber(context.Background())
 		if err != nil {
+			a.metrics.BackendErrors.Inc()
```
Comment (istae) [inline pkg/storageincentives/agent.go:246]: nice

**Review by istae — CHANGES_REQUESTED** [review]
Nice work, small suggestions :)

File: pkg/storageincentives/metrics.go [hunk 7]
```diff
@@ -72,6 +84,52 @@ func newMetrics() metrics {
...
+		ErrWinner: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: m.Namespace,
+			Subsystem: subsystem,
+			Name:      "winPhase_errors",
+			Help:      "total win phase while processing",
+		}),
+		ErrCheckIsPlaying: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: m.Namespace,
+			Subsystem: subsystem,
+			Name:      "neighborhoodSelected_errors",
```
Comment (istae) [inline pkg/storageincentives/metrics.go:130]: rename to is_playing_errors

File: pkg/storageincentives/agent.go [hunk 8]
```diff
@@ -119,8 +119,10 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		mtx.Unlock()
 
 		if round-1 == sampleRound { // the sample has to come from previous round to be able to commit it
+			a.metrics.CommitPhase.Inc()
 			obf, err := a.commit(ctx, storageRadius, reserveSample)
 			if err != nil {
+				a.metrics.ErrCommit.Inc()
```
Comment (vladopajic) [inline pkg/storageincentives/agent.go:125]: Could `CommitPhase` and `ErrCommit` be calculated in `a.commit` func?

File: pkg/storageincentives/agent.go [hunk 9]
```diff
@@ -160,8 +162,10 @@ func (a *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase ui
 		mtx.Unlock()
 
 		if round == commitRound { // reveal requires the obfuscationKey from the same round
+			a.metrics.RevealPhase.Inc()
 			err := a.reveal(ctx, storageRadius, reserveSample, obfuscationKey)
 			if err != nil {
+				a.metrics.ErrReveal.Inc()
 				a.logger.Error(err, "reveal")
```
Comment (vladopajic) [inline pkg/storageincentives/agent.go:169]: Could `RevealPhase` and `ErrReveal` be calculated in `a.reveal` func?

File: pkg/storageincentives/metrics.go [hunk 10]
```diff
@@ -18,6 +19,17 @@ type metrics struct {
 	NeighborhoodSelected prometheus.Counter
 	SampleDuration       prometheus.Gauge
 	Round                prometheus.Gauge
+
+	// total calls to chain backend
+	BackendCalls  prometheus.Counter
+	BackendErrors prometheus.Counter
```
Comment (vladopajic) [inline pkg/storageincentives/metrics.go:25]: Ideally these should be exposed by `ChainBackend` interface (or any other type wrapping this interface in order to provide this this functionality.

**Comment by vladopajic** [comment]
Ideally metrics should be defined and calculated inside the same scope where logic (which is being measured) is defined. 

That is, if measured function is `doTask()` it would be the best to put metrics inside this function 
```
func doTask() (err error) {
	a.metrics.DoTaskCalls().Inc()
	defer func() {
		if err {
			a.metrcis.DoCallsErr().inc()
		}
	}()

	// ... 
}
```

Having this practice enables more robust code because code using `doTask()` function will not have to take care of calculating it's metrics; also it improves code reuse.  For example `BackendCalls` should be measured inside this type, and if it is implemented this way then all other logic would have these metrics out of the box.

**Comment by swarmHaseeb** [comment]
> Ideally metrics should be defined and calculated inside the same scope where logic (which is being measured) is defined.
> 
> That is, if measured function is `doTask()` it would be the best to put metrics inside this function
> 
> ```
> func doTask() (err error) {
> 	a.metrics.DoTaskCalls().Inc()
> 	defer func() {
> 		if err {
> 			a.metrcis.DoCallsErr().inc()
> 		}
> 	}()
> 
> 	// ... 
> }
> ```
> 
> Having this practice enables more robust code because code using `doTask()` function will not have to take care of calculating it's metrics; also it improves code reuse. For example `BackendCalls` should be measured inside this type, and if it is implemented this way then all other logic would have these metrics out of the box.

makes sense, thankyou, will keep it in mind.

**Review by istae — APPROVED** [review]

**Pull request merged**